tools/check_format.py

```diff
@@ -376,6 +449,11 @@ def checkFormatVisitor(arg, dir_name, names):
   if args.add_excluded_prefixes:
     EXCLUDED_PREFIXES += tuple(args.add_excluded_prefixes)
 
+  # Check whether all needed external tools are available.
+  ct_error_messages = checkTools()
+  if checkErrorMessages(ct_error_messages):
```

Comment:
What is the motivation to continue? The check format will fail at the end anyway, and it will print out bunch of `command not found` error burying the real error messages. Setting up tools is not hard (esp. if you use docker) and IMO we should fail here when run locally.

Reply:
sure it's not a big deal either way I guess. Just a preference that the tool should tell me what's wrong to the best of it's abilities in general.

Picture this is happening in CI -- I realize this is a bit contrived as CI should have all the tools. It would be better in principle if it told you all it could about what was wrong so you could go and fix it without waiting for another push and waiting for your turn.

Of course when you are running locally it's easy to iterate.

Reply:
Alright, I think we should merge this since @mrostecki is not a regular contributor, might not have more time to work on this, and I think this is a clear improvement. We can do further follow ups if we want.

Reply:
Stopping there is actually a reason why I submitted this PR.

If you don't have all tools locally, you install git hooks and try to push something, then you end up with thousands of `command not found` lines on the console. ctrl+c does not work, because all the checks below are running in multiple subprocesses which don't have any special signal handling. So, you end up with seing `command not found` messages for at least two minutes without being able to do anything about that.

If tools are missing, it's obvious that checks are going to fail. I don't understand why seing more errors except those about the lack of tools would change anything.

Sorry for the silence! I was quite busy last days.

Feel free to share thoughts/comments if you disagree with what I wrote. And I'm open for doing follow up PR-s if we agree that something needs to be changed.

Reply:
no worries; leave it like it is. A comment though why the early exit is warranted here would help, as I want the pattern in general to be that we find all the errors and print them, so exceptions should be called out.

I think check_format.py can still do most of what it is trying to do without external tools, but if clang is not present I can see how getting an error message for each file would be a drag. Those are run in parallel in background threads so are probably hard to interrupt. But given your experience I agree with your conclusion.

Reply:
@jmarantz Another drawback of continuing is that `clang-format fix` file operation is not atomic. I experienced a couple times that I want to interrupt `tools/check_format.py fix` with Ctrl-C because buildifier is not found, then it ended up with deleting some lines of source code that clang-format is fixing on the fly and it was hard to recover if changes aren't committed. so perhaps we can put this behavior behind a flag.

Reply:
@lizan that sounds unrelated to this, which is about the 'check' operation showing you all the issues. I too have experienced sadness while attempting to kill an in-progress 'fix'. Can you open up a new bug about handing ^C gracefully for 'fix'?

Reply:
@jmarantz IIUC, the check added here will stop invoking clang-format or buildifier if tools are not found for `fix` flow as well. So it would prevent the case I would need to do ^C when I found some command is not available.

Reply:
I agree that the early-exit is fine in this case, and I just ask for a comment explaining why.